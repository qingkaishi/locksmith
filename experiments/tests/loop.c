#include <pthread.h>
#include <stdio.h>

int *j;

pthread_mutex_t l;

void * dostuff(void * a) {
  pthread_mutex_lock(&l);
  j = 0;
  pthread_mutex_unlock(&l);
  return NULL;
}

int main() {
  pthread_t t1, t2;

  int i = 1;
  while(1) {
    pthread_mutex_init(&l, NULL);  // makes lock non-linear
  }
  pthread_create(&t1, NULL, dostuff, NULL);
  pthread_create(&t2, NULL, dostuff, NULL);
  return 1;
}

