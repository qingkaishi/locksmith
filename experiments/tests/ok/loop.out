************************* STARTING *************************

shared: &j:loop.c:4

Warning: Possible data race: &j:loop.c:4 is protected by non-linear or concrete lock(s):
  l:loop.c:6
  non linear concrete l:loop.c:20
 references:
  dereference of &j:loop.c:4 at loop.c:10
    &j:loop.c:4
  locks acquired:
    l:loop.c:6
    non linear concrete l:loop.c:20
  in: main at loop.c:15 


*************************** DONE ***************************
